flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/HsSelectiveSpillingStrategy.java

```diff
@@ -0,0 +1,122 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.io.network.partition.hybrid;
+
+import java.util.ArrayDeque;
+import java.util.ArrayList;
+import java.util.Deque;
+import java.util.List;
+import java.util.PriorityQueue;
+
+import static org.apache.flink.util.Preconditions.checkNotNull;
+
+/**
+ * A special implementation of {@link HsSpillingStrategy} that reduce disk writes as much as
+ * possible.
+ */
+public class HsSelectiveSpillingStrategy implements HsSpillingStrategy {
+    public final float spillBufferRatio;
+
+    public final float spillThreshold;
+
+    public HsSelectiveSpillingStrategy(HybridShuffleConfiguration hybridShuffleConfiguration) {
+        spillThreshold = hybridShuffleConfiguration.getSelectiveSpillThreshold();
+        spillBufferRatio = hybridShuffleConfiguration.getSelectiveSpillBufferRatio();
+    }
+
+    /**
+     * For the case of buffer finished, there is no need to make a decision for {@link
+     * HsSelectiveSpillingStrategy}.
+     */
+    @Override
+    public Action onBufferFinished(
+            HsSpillingInfoProvider spillingInfoProvider, BufferWithIdentity finishedBuffer) {
+        return Action.EMPTY;
+    }
+
+    /**
+     * For the case of buffer consumed, this buffer need release. The control of the buffer is taken
+     * over by the downstream task.
+     */
+    @Override
+    public Action onBufferConsumed(
+            HsSpillingInfoProvider spillingInfoProvider, BufferWithIdentity consumedBuffer) {
+        return Action.builder().addBufferToRelease(consumedBuffer).build();
+    }
+
+    /**
+     * When the amount of memory used exceeds the threshold, trigger the spilling operation and
+     * score the buffer of each subpartition. The lower the score, the more likely the buffer will
+     * be consumed in the next time, and should be kept in memory as much as possible. Select all
+     * buffers that need to be spilled according to the score from high to low.
+     */
+    @Override
+    public Action checkForAction(HsSpillingInfoProvider spillInfoProvider) {
+        if (spillInfoProvider.getNumTotalUnSpilledBuffers()
+                < spillInfoProvider.getPoolSize() * spillThreshold) {
+            return Action.EMPTY;
+        }
+
+        List<Integer> consumptionProgress = spillInfoProvider.getConsumptionProgress();
+        PriorityQueue<BufferScoreQueue> heap = new PriorityQueue<>();
+        for (int channel = 0; channel < spillInfoProvider.getNumSubpartitions(); channel++) {
+            Deque<BufferWithIdentity> unSpilledBuffers =
+                    spillInfoProvider.getUnSpilledBuffers(channel);
+            // In order to prevent the impact on the original queue, copy it and then pass to heap.
+            Deque<BufferWithIdentity> copiedQueue = new ArrayDeque<>(unSpilledBuffers);
```

Comment:
Instead of copying the deque, we may consider using an iterator.

Reply:
Rewritten here using PeekingIterator